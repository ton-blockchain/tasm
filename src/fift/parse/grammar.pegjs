{{}}

SourceFile
  = space* include:IncludeDirective? space* program:Program space* {
    return { $: 'SourceFile', include, program };
  }

IncludeDirective
  = "\"" path:$[^"]+ "\"" space+ "include" space* {
    return { $: 'IncludeDirective', path };
  }

Program
  = "PROGRAM{" space* declarations:Declaration* space* definitions:Definition* space* "}END>c" {
    return { $: 'Program', declarations, definitions };
  }

Declaration
  = space* decl:(ProcDeclaration / MethodDeclaration / GlobalVar) space* {
    return { $: 'Declaration', decl };
  }

ProcDeclaration
  = "DECLPROC" space+ name:Identifier {
    return { $: 'ProcDeclaration', name };
  }

MethodDeclaration
  = method_id:Integer space+ "DECLMETHOD" space+ name:Identifier {
    return { $: 'MethodDeclaration', method_id, name };
  }

GlobalVar
  = "DECLGLOBVAR" space+ name:Identifier {
    return { $: 'GlobalVar', name };
  }

Definition
  = space* def:(ProcDefinition / ProcInlineDefinition / ProcRefDefinition / MethodDefinition) space* {
    return { $: 'Definition', def };
  }

ProcDefinition
  = name:Identifier space* "PROC:<{" space* instructions:Instruction* "}>" {
    return { $: 'ProcDefinition', name, instructions };
  }

ProcInlineDefinition
  = name:Identifier space* "PROCINLINE:<{" space* instructions:Instruction* "}>" {
    return { $: 'ProcInlineDefinition', name, instructions };
  }

ProcRefDefinition
  = name:Identifier space* "PROCREF:<{" space* instructions:Instruction* "}>" {
    return { $: 'ProcRefDefinition', name, instructions };
  }

MethodDefinition
  = name:Identifier space* "METHOD:<{" space* instructions:Instruction* "}>" {
    return { $: 'MethodDefinition', name, instructions };
  }

Instruction
  = space* instr:(IfStatement / IfjmpStatement / WhileStatement / RepeatStatement / UntilStatement / AsmExpression) space* {
    return { $: 'Instruction', instr };
  }

AsmExpression
  = args:AsmArgumentList? space* name:TvmInstruction {
    return { $: 'AsmExpression', arguments: args, name };
  }

AsmArgumentList
  = space* primitives:AsmPrimitive+ {
    return { $: 'AsmArgumentList', primitives };
  }

AsmPrimitive
  = space* prim:(InstructionBlock / String / HexBitString / BinBitString / BocHex / FiftAddressNone / StackRegister / ControlRegister / Integer / ArgIdentifier) space* {
    return { $: 'AsmPrimitive', prim };
  }

IfStatement
  = kind:("IF:<{" / "IFNOT:<{") space* instructions:Instruction* "}>" else_block:("ELSE<{" space* else_instructions:Instruction* "}>" { return { instructions: else_instructions }; })? {
    return { $: 'IfStatement', kind, instructions, else_block: else_block ?? undefined };
  }

IfjmpStatement
  = kind:("IFJMP:<{" / "IFNOTJMP:<{") space* instructions:Instruction* "}>" {
    return { $: 'IfjmpStatement', kind, instructions };
  }

WhileStatement
  = "WHILE:<{" space* condition:Instruction* "}>DO<{" space* body:Instruction* "}>" {
    return { $: 'WhileStatement', condition, body };
  }

RepeatStatement
  = "REPEAT:<{" space* instructions:Instruction* "}>" {
    return { $: 'RepeatStatement', instructions };
  }

UntilStatement
  = "UNTIL:<{" space* instructions:Instruction* "}>" {
    return { $: 'UntilStatement', instructions };
  }

InstructionBlock
  = "<{" space* instructions:Instruction* ("}>c" / "}>s" / "}>CONT" / "}>") {
    return { $: 'InstructionBlock', instructions };
  }

HexBitString
  = "x{" content:$([0-9a-fA-F]* "_"?) "}" {
    return { $: 'HexBitString', content };
  }

BinBitString
  = "b{" content:$[01]* "}" {
    return { $: 'BinBitString', content };
  }

BocHex
  = "B{" content:$[0-9a-fA-F_]+ "}" {
    return { $: 'BocHex', content };
  }

FiftAddressNone
  = "<b" space* "0" space* "2" space* "u," space* "b>" space* "<s" {
    return { $: 'FiftAddressNone' };
  }

Identifier
  = name:$(!reservedWord [a-zA-Z~$_%?.] [a-zA-Z0-9$_%?()~.]*) {
    return { $: 'Identifier', name };
  }

ArgIdentifier
  = name:$(!reservedWord (([A-Z] [a-z]) / [a-z$_%?~.]) [a-zA-Z0-9$_%?()~.]*) {
    return { $: 'ArgIdentifier', name };
  }

TvmInstruction
  = instr:$("-"? [A-Z0-9_#:] [A-Z0-9_#:]+ "l"?) {
    return { $: 'TvmInstruction', value: instr };
  }

Integer
  = value:$("-"? "0b" [01]+) {
    return { $: 'Integer', value };
  }
  / value:$("-"? "0x" [0-9a-fA-F]+) {
    return { $: 'Integer', value };
  }
  / value:$("-"? [0-9]+) {
    return { $: 'Integer', value };
  }

StackRegister
  = value:$("s" [0-9][0-9]?) {
    return { $: 'StackRegister', value };
  }
  / value:$("s(" "-"?[0-9][0-9]? ")") {
    return { $: 'StackRegister', value };
  }
  / value:$([0-9][0-9]?[0-9]? "s()") {
    return { $: 'StackRegister', value };
  }

ControlRegister
  = value:$("c" [0-9][0-9]?) {
    return { $: 'ControlRegister', value };
  }

String
  = "\"" content:$[^"]* "\"" {
    return { $: 'String', content };
  }

reservedWord
  = "PROGRAM" !idChar
  / "END>c" !idChar
  / "DECLPROC" !idChar
  / "DECLMETHOD" !idChar
  / "DECLGLOBVAR" !idChar
  / "PROC:<{" 
  / "PROCINLINE:<{"
  / "PROCREF:<{"
  / "METHOD:<{"
  / "IF:<{"
  / "IFNOT:<{"
  / "ELSE<{"
  / "IFJMP:<{"
  / "IFNOTJMP:<{"
  / "WHILE:<{"
  / "DO<{"
  / "REPEAT:<{"
  / "UNTIL:<{"
  / "<b 0 2 u, b>"
  / "CALLDICT" !idChar
  / "INLINECALLDICT" !idChar

idChar
  = [a-zA-Z0-9$_%?]

space
  = [ \t\r\n] / comment

comment
  = "//" [^\r\n]*
