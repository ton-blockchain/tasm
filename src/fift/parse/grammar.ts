/* eslint-disable @typescript-eslint/no-namespace */
/* eslint-disable @typescript-eslint/ban-types */
/* eslint-disable @typescript-eslint/no-redundant-type-constituents */
/* eslint-disable @typescript-eslint/no-duplicate-type-constituents */
/* eslint-disable @typescript-eslint/no-unused-vars */
export namespace $ast {
    export type SourceFile = {
        readonly $: "SourceFile"
        readonly include: IncludeDirective | undefined
        readonly program: Program
    }

    export type IncludeDirective = {
        readonly $: "IncludeDirective"
        readonly path: string
    }

    export type Program = {
        readonly $: "Program"
        readonly declarations: readonly Declaration[]
        readonly definitions: readonly Definition[]
    }

    export type Declaration = {
        readonly $: "Declaration"
        readonly decl: ProcDeclaration | MethodDeclaration | GlobalVar
    }

    export type ProcDeclaration = {
        readonly $: "ProcDeclaration"
        readonly name: Identifier
    }

    export type MethodDeclaration = {
        readonly $: "MethodDeclaration"
        readonly method_id: Integer
        readonly name: Identifier
    }

    export type GlobalVar = {
        readonly $: "GlobalVar"
        readonly name: Identifier
    }

    export type Definition = {
        readonly $: "Definition"
        readonly def: ProcDefinition | ProcInlineDefinition | ProcRefDefinition | MethodDefinition
    }

    export type ProcDefinition = {
        readonly $: "ProcDefinition"
        readonly name: Identifier
        readonly instructions: readonly Instruction[]
    }

    export type ProcInlineDefinition = {
        readonly $: "ProcInlineDefinition"
        readonly name: Identifier
        readonly instructions: readonly Instruction[]
    }

    export type ProcRefDefinition = {
        readonly $: "ProcRefDefinition"
        readonly name: Identifier
        readonly instructions: readonly Instruction[]
    }

    export type MethodDefinition = {
        readonly $: "MethodDefinition"
        readonly name: Identifier
        readonly instructions: readonly Instruction[]
    }

    export type Instruction = {
        readonly $: "Instruction"
        readonly instr:
            | IfStatement
            | IfjmpStatement
            | WhileStatement
            | RepeatStatement
            | UntilStatement
            | AsmExpression
    }

    export type AsmExpression = {
        readonly $: "AsmExpression"
        readonly arguments: AsmArgumentList | undefined
        readonly name: TvmInstruction
    }

    export type AsmArgumentList = {
        readonly $: "AsmArgumentList"
        readonly primitives: readonly AsmPrimitive[]
    }

    export type AsmPrimitive = {
        readonly $: "AsmPrimitive"
        readonly prim:
            | InstructionBlock
            | String
            | HexBitString
            | BinBitString
            | BocHex
            | StackRegister
            | ControlRegister
            | Integer
            | ArgIdentifier
            | FiftAddressNone
    }

    export type IfStatement = {
        readonly $: "IfStatement"
        readonly kind: "IF:<{" | "IFNOT:<{"
        readonly instructions: readonly Instruction[]
        readonly else_block:
            | {
                  readonly instructions: readonly Instruction[]
              }
            | undefined
    }

    export type IfjmpStatement = {
        readonly $: "IfjmpStatement"
        readonly kind: "IFJMP:<{" | "IFNOTJMP:<{"
        readonly instructions: readonly Instruction[]
    }

    export type WhileStatement = {
        readonly $: "WhileStatement"
        readonly condition: readonly Instruction[]
        readonly body: readonly Instruction[]
    }

    export type RepeatStatement = {
        readonly $: "RepeatStatement"
        readonly instructions: readonly Instruction[]
    }

    export type UntilStatement = {
        readonly $: "UntilStatement"
        readonly instructions: readonly Instruction[]
    }

    export type InstructionBlock = {
        readonly $: "InstructionBlock"
        readonly instructions: readonly Instruction[]
    }

    export type HexBitString = {
        readonly $: "HexBitString"
        readonly content: string
    }

    export type BinBitString = {
        readonly $: "BinBitString"
        readonly content: string
    }

    export type FiftAddressNone = {
        readonly $: "FiftAddressNone"
    }

    export type BocHex = {
        readonly $: "BocHex"
        readonly content: string
    }

    export type Identifier = {
        readonly $: "Identifier"
        readonly name: string
    }

    export type ArgIdentifier = {
        readonly $: "ArgIdentifier"
        readonly name: string
    }

    export type TvmInstruction = {
        readonly $: "TvmInstruction"
        readonly value: string
    }

    export type Integer = {
        readonly $: "Integer"
        readonly value: string
    }

    export type StackRegister = {
        readonly $: "StackRegister"
        readonly value: string
    }

    export type ControlRegister = {
        readonly $: "ControlRegister"
        readonly value: string
    }

    export type String = {
        readonly $: "String"
        readonly content: string
    }
}
