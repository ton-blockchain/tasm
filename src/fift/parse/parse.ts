import type * as G from "./grammar"
import * as $ from "./grammar.gen.pegjs"

export type ParseResult = ParseSuccess | ParseFailure

export type ParseSuccess = {
    readonly $: "ParseSuccess"
    readonly ast: G.$ast.SourceFile
}

export type ParseFailure = {
    readonly $: "ParseFailure"
    readonly error: ParseError
}

export class ParseError extends Error {
    public readonly position?: PegLocation

    public constructor(message: string, position?: PegLocation) {
        super(message)
        this.name = "ParseError"
        this.position = position
    }
}

const success = (ast: G.$ast.SourceFile): ParseSuccess => ({$: "ParseSuccess", ast})
const failure = (error: ParseError): ParseFailure => ({$: "ParseFailure", error})

export type PegPosition = {
    readonly offset: number
    readonly line: number
    readonly column: number
}

export type PegLocation = {
    readonly source?: string
    readonly start: PegPosition
    readonly end: PegPosition
}

export type PegParseError = {
    readonly message: string
    readonly location: PegLocation
}

export function parse(_filepath: string, code: string): ParseResult {
    try {
        const normalizedCode = code.replaceAll("2DROP", "DROP2").replaceAll("}>CONT", "}> PUSHCONT")
        const ast = $.parse(normalizedCode, {
            startRule: "SourceFile",
        }) as G.$ast.SourceFile
        return success(ast)
    } catch (error) {
        if (error instanceof Error) {
            const pegError = error as unknown as PegParseError
            const message = pegError.message.slice(0, -1) // trim trailing `.`
            return failure(new ParseError(message, pegError.location))
        }
        return failure(new ParseError("Unknown parse error"))
    }
}
