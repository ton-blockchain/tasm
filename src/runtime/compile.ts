import * as c from "./constructors"
import * as $ from "./util"
import type {Instr} from "./instr-gen"
import {CodeBuilder} from "./builder"
import {instr} from "./instr"
import {matchingRule} from "./layout"

export const compileInstructions: $.Store<Instr[]> = (b: CodeBuilder, instructions: Instr[]) => {
    for (let index = 0; index < instructions.length; index++) {
        const instruction = instructions[index]
        if (!instruction) break
        const builderBefore = new CodeBuilder().storeBuilder(b)

        const overflow = safeStore(b, instruction)
        if (!overflow) {
            // fast path
            continue
        }

        // Handle case where instructions don't fit to the current cell.
        // In many cases, this happens at the if-else boundary, when the body
        // specified via PUSHCONT takes up a large number of bits. In this case, we
        // need to create a new ref and continue compiling the remaining instructions
        // into this ref
        const remainingInstructions = instructions.slice(index)

        // But we can try to optimize code like this:
        // ref {
        //   PUSHCONT { ... }
        //   IF
        //   ...
        // }
        // ->
        // IFREF { ... }
        // ...
        const match = matchingRule(remainingInstructions)
        if (match) {
            const instr = match.rule.ctor(match.body)
            match.rule.type.store(builderBefore, instr)
            b.reinitFrom(builderBefore)
            index++ // advance to not store instruction once again

            // All instructions after PUSHCONT + IF will be compiled in the current cell and
            // placed in a new cell if necessary.
            continue
        }

        // Create a new ref and compile the remaining instruction to it
        $.PSEUDO_PUSHREF.store(builderBefore, c.PSEUDO_PUSHREF($.code(remainingInstructions)))
        b.reinitFrom(builderBefore)
        // All remaining instructions already processed in PSEUDO_PUSHREF,
        // so we need to return here
        return
    }
}

const safeStore = (b: CodeBuilder, t: Instr): boolean => {
    try {
        instr.store(b, t)
        if (b.bits >= 1023) {
            return true
        }

        if (b.refs >= 4) {
            if (t.$ === "PSEUDO_PUSHREF" && b.refs === 4) {
                // In the case where the compiler itself has set the necessary `ref {}`,
                // we do not try to predict in advance whether there may be an overflow further,
                // since the compiler has already decomposed everything into correct refs
                return false
            }
            // In case of other instructions that push references, we cannot be sure, s
            // o we handle this case explicitly.
            return true
        }
    } catch (error: unknown) {
        if (error instanceof Error) {
            if (
                error.message === "BitBuilder overflow" ||
                error.message === "Too many references"
            ) {
                return true
            }
        }

        throw error
    }

    return false
}
