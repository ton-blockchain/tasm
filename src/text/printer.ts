import type {Instr} from "../runtime"
import type {Code, Dict} from "../runtime/util"
import type {Slice} from "@ton/core"
import {printInstruction} from "./printer-gen"

export class Printer {
    public indent: number
    public result: string

    public constructor(indent: number) {
        this.indent = indent
        this.result = ""
    }

    public beginLine(text: string) {
        const indentText = "    ".repeat(this.indent)
        this.result += indentText + text
    }

    public append(text: string) {
        this.result += text
    }

    public inIndent(action: (p: Printer) => void) {
        this.indent++
        action(this)
        this.indent--
    }

    public build(): string {
        return this.result
    }
}

export const printInstr = (p: Printer, instr: Instr) => {
    if (instr.$ === "PSEUDO_PUSHREF") {
        p.beginLine("ref ")
        printCode(p, instr.arg0)
        return
    }

    if (instr.$ === "PSEUDO_PUSHSLICE") {
        p.beginLine("embed ")
        printSlice(p, instr.arg0)
        return
    }

    if (instr.$ === "PSEUDO_EXOTIC") {
        p.beginLine("exotic ")

        if (instr.arg0.$ === "LibraryCell") {
            p.append("library ")
            printSlice(p, instr.arg0.data)
            return
        }

        printSlice(p, instr.arg0.cell.beginParse(true))
        return
    }

    printInstruction(p, instr)
}

export const printInstructions = (p: Printer, instructions: Instr[]) => {
    if (instructions.length === 0) {
        p.append("{}")
        return
    }

    p.append("{\n")
    p.inIndent(p => {
        for (const instr of instructions) {
            printInstr(p, instr)
            p.append("\n")
        }
    })
    p.beginLine("}")
}

export const print = (instructions: Instr[]) => {
    const p = new Printer(0)
    for (const instr of instructions) {
        printInstr(p, instr)
        p.append("\n")
    }
    return p.build()
}

export function printDictionary(p: Printer, dict: Dict) {
    if (dict.$ === "RawDict") {
        printSlice(p, dict.slice)
    }

    if (dict.$ === "DecompiledDict") {
        p.append(`[\n`)

        for (const method of dict.methods) {
            p.inIndent(p => {
                p.beginLine(`${method.id} => `)
                printInstructions(p, method.instructions)
                p.append(`\n`)
            })
        }

        p.beginLine("]")
    }
}

export function printCode(p: Printer, arg0: Code) {
    if (arg0.$ === "Instructions") {
        printInstructions(p, arg0.instructions)
    }

    if (arg0.$ === "Raw") {
        printSlice(p, arg0.slice)
    }
}

export function printSlice(p: Printer, slice: Slice) {
    const hasRefs = slice.remainingRefs
    if (hasRefs) {
        const data = slice.asCell().toBoc().toString("hex")
        p.append(`boc{${data}}`)
    } else {
        const data = slice.asCell().bits.toString()
        p.append(`x{${data}}`)
    }
}
