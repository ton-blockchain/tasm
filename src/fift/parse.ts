import * as $ from "@tonstudio/parser-runtime"
import * as G from "./grammar"

export type ParseResult = ParseSuccess | ParseFailure

export type ParseSuccess = {
    readonly $: "ParseSuccess"
    readonly ast: G.$ast.SourceFile
}

export type ParseFailure = {
    readonly $: "ParseFailure"
    readonly error: ParseError
}

export class ParseError extends Error {
    public readonly position?: number

    public constructor(message: string, position?: number) {
        super(message)
        this.name = "ParseError"
        this.position = position
    }
}

const success = (ast: G.$ast.SourceFile): ParseSuccess => ({$: "ParseSuccess", ast})
const failure = (error: ParseError): ParseFailure => ({$: "ParseFailure", error})

export function parse(filepath: string, code: string): ParseResult {
    const res = $.parse({
        grammar: G.SourceFile,
        space: G.space,
        text: code,
    })

    if (res.$ !== "success") {
        const {expected, position} = res.error
        return failure(
            new ParseError(
                `Expected ${getExpectedText(expected)} at position ${position}`,
                position,
            ),
        )
    }

    return success(res.value)
}

const getExpectedText = (expected: ReadonlySet<string>) => {
    const result: string[] = []
    const failures = [...expected].sort()
    for (const [idx, failure] of failures.entries()) {
        if (idx > 0) {
            if (idx === failures.length - 1) {
                result.push(failures.length > 2 ? ", or " : " or ")
            } else {
                result.push(", ")
            }
        }
        result.push(failure)
    }
    return result.join("")
}
