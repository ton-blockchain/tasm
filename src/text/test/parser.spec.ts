import {decompileCell} from "../../runtime"
import {print} from "../printer"
import {readFileSync} from "node:fs"
import {parse} from "../parse"
import {boc} from "../../runtime/util"

describe("assembly-parser", () => {
    it("should parse simple assembly", () => {
        const code = `
            PUSHINT_4 10
            PUSHINT_4 5
            ADD
        `
        const res = parse("test.asm", code)
        if (res.$ === "ParseFailure") {
            throw new Error("unexpected parser error")
        }

        expect(print(res.instructions)).toMatchSnapshot()
    })

    it("should parse assembly with raw pushref", () => {
        const code = `
            PUSHREF x{71}
        `
        const res = parse("test.asm", code)
        if (res.$ === "ParseFailure") {
            throw new Error("unexpected parser error")
        }

        expect(print(res.instructions)).toMatchSnapshot()
    })

    it("should parse assembly with invalid raw pushref", () => {
        const code = `
            PUSHREF x{22221}
        `
        const res = parse("test.asm", code)
        if (res.$ === "ParseFailure") {
            throw new Error("unexpected parser error")
        }

        expect(print(res.instructions)).toMatchSnapshot()
    })

    it("should parse and print assembly", () => {
        const instructions = decompileCell(
            boc(
                readFileSync(
                    `${__dirname}/testdata/jetton_minter_discoverable_JettonMinter.boc`,
                ).toString("hex"),
            ).asCell(),
        )
        const assembly = print(instructions)

        const res = parse("test.asm", assembly)
        if (res.$ === "ParseFailure") {
            throw new Error("unexpected parser error")
        }

        const assembly2 = print(res.instructions)

        expect(assembly2).toEqual(assembly)
    })

    it("should not parse assembly with error", () => {
        const code = `
            PUSHINT_4 10 ,
            PUSHINT_4 5
            ADD
        `
        const res = parse("test.asm", code)
        if (res.$ === "ParseSuccess") {
            throw new Error("unexpected parser success")
        }

        expect(res.error.toString()).toMatchSnapshot()
    })

    it("should give an error for malformed assembly", () => {
        const code = `
            PUSHINT_4 // no arg
            PUSHINT_4 5
            ADD
        `
        const res = parse("test.asm", code)
        if (res.$ === "ParseSuccess") {
            throw new Error("unexpected parser success")
        }

        expect(res.error.toString()).toMatchSnapshot()
    })

    it("should parse big hex number", () => {
        const code = `
            PUSHINT_LONG 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
            PUSHINT_LONG -0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
            ADD
        `
        const res = parse("test.asm", code)
        if (res.$ === "ParseFailure") {
            throw new Error("unexpected parser error")
        }

        expect(print(res.instructions)).toMatchSnapshot()
    })

    it("should parse big decimal number", () => {
        const code = `
            PUSHINT_LONG 999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
            PUSHINT_LONG -999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
            ADD
        `
        const res = parse("test.asm", code)
        if (res.$ === "ParseFailure") {
            throw new Error("unexpected parser error")
        }

        expect(print(res.instructions)).toMatchSnapshot()
    })

    it("should parse big binary number", () => {
        const code = `
            PUSHINT_LONG 0b111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
            PUSHINT_LONG -0b111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
            ADD
        `
        const res = parse("test.asm", code)
        if (res.$ === "ParseFailure") {
            throw new Error("unexpected parser error")
        }

        expect(print(res.instructions)).toMatchSnapshot()
    })

    it("should parse big octal number", () => {
        const code = `
            PUSHINT_LONG 0o7777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777
            PUSHINT_LONG -0o7777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777777
            ADD
        `
        const res = parse("test.asm", code)
        if (res.$ === "ParseFailure") {
            throw new Error("unexpected parser error")
        }

        expect(print(res.instructions)).toMatchSnapshot()
    })

    it("should parse hex number", () => {
        const code = `
            PUSHINT_16 0xFFF
            PUSHINT_16 -0xFFF
            ADD
        `
        const res = parse("test.asm", code)
        if (res.$ === "ParseFailure") {
            throw new Error("unexpected parser error")
        }

        expect(print(res.instructions)).toMatchSnapshot()
    })

    it("should parse decimal number", () => {
        const code = `
            PUSHINT_16 999999
            PUSHINT_16 -999999
            ADD
        `
        const res = parse("test.asm", code)
        if (res.$ === "ParseFailure") {
            throw new Error("unexpected parser error")
        }

        expect(print(res.instructions)).toMatchSnapshot()
    })

    it("should parse binary number", () => {
        const code = `
            PUSHINT_16 0b1111111111
            PUSHINT_16 -0b1111111111
            ADD
        `
        const res = parse("test.asm", code)
        if (res.$ === "ParseFailure") {
            throw new Error("unexpected parser error")
        }

        expect(print(res.instructions)).toMatchSnapshot()
    })

    it("should parse octal number", () => {
        const code = `
            PUSHINT_16 0o77777
            PUSHINT_16 -0o77777
            ADD
        `
        const res = parse("test.asm", code)
        if (res.$ === "ParseFailure") {
            throw new Error("unexpected parser error")
        }

        expect(print(res.instructions)).toMatchSnapshot()
    })

    it("should give an error for too big number ", () => {
        const code = `
            PUSHINT_4 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
            PUSHINT_4 5
            ADD
        `
        const res = parse("test.asm", code)
        if (res.$ === "ParseSuccess") {
            throw new Error("unexpected parser success")
        }

        expect(res.error.toString()).toMatchSnapshot()
    })

    it("should parse hex number with _", () => {
        const code = `
            PUSHINT_16 0xF_FF
            PUSHINT_16 -0xF_F_F
            ADD
        `
        const res = parse("test.asm", code)
        if (res.$ === "ParseFailure") {
            throw new Error(`unexpected parser error ${res.error.message}`)
        }

        expect(print(res.instructions)).toMatchSnapshot()
    })

    it("should parse decimal number with _", () => {
        const code = `
            PUSHINT_16 999_999
            PUSHINT_16 -99_99_99
            ADD
        `
        const res = parse("test.asm", code)
        if (res.$ === "ParseFailure") {
            throw new Error("unexpected parser error")
        }

        expect(print(res.instructions)).toMatchSnapshot()
    })

    it("should parse binary number with _", () => {
        const code = `
            PUSHINT_16 0b11111_11111
            PUSHINT_16 -0b111_11_11_1_11
            ADD
        `
        const res = parse("test.asm", code)
        if (res.$ === "ParseFailure") {
            throw new Error("unexpected parser error")
        }

        expect(print(res.instructions)).toMatchSnapshot()
    })

    it("should parse octal number with _", () => {
        const code = `
            PUSHINT_16 0o777_77
            PUSHINT_16 -0o7_77_77
            ADD
        `
        const res = parse("test.asm", code)
        if (res.$ === "ParseFailure") {
            throw new Error("unexpected parser error")
        }

        expect(print(res.instructions)).toMatchSnapshot()
    })
})
