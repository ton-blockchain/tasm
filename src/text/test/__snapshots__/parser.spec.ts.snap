// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assembly-parser should give an error for malformed assembly 1`] = `"ParseError: Expected 1 argument at test.asm:1"`;

exports[`assembly-parser should give an error for too big number  1`] = `"ParseError: Number argument is too big, max value is 9007199254740991, but 2707685248164858261307045101702230179137145581421695874189921465443966120903931272499975005961073806735733604454495675614232575 given at test.asm:1"`;

exports[`assembly-parser should not parse assembly with error 1`] = `"ParseError: Expected "-", "0", "[", "\\"", "boc{", "b{", "c", "embed", "exotic", "ref", "s", "x{", "{", digit, or identifier at test.asm:1"`;

exports[`assembly-parser should parse assembly with invalid raw pushref 1`] = `
"PUSHREF {
    embed x{22221}
}
"
`;

exports[`assembly-parser should parse assembly with raw pushref 1`] = `
"PUSHREF {
    PUSHINT_4 1
}
"
`;

exports[`assembly-parser should parse big binary number 1`] = `
"PUSHINT_LONG 99361157907245371849534687261600163536440705095468583112899330433667260971928120725176218165033374588831214764616388685194981122823348121052434385602635592909085118465334393955216978280422300468816831948397854000057142198644227702783
PUSHINT_LONG -99361157907245371849534687261600163536440705095468583112899330433667260971928120725176218165033374588831214764616388685194981122823348121052434385602635592909085118465334393955216978280422300468816831948397854000057142198644227702783
ADD
"
`;

exports[`assembly-parser should parse big decimal number 1`] = `
"PUSHINT_LONG 999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
PUSHINT_LONG -999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999
ADD
"
`;

exports[`assembly-parser should parse big hex number 1`] = `
"PUSHINT_LONG 2707685248164858261307045101702230179137145581421695874189921465443966120903931272499975005961073806735733604454495675614232575
PUSHINT_LONG -2707685248164858261307045101702230179137145581421695874189921465443966120903931272499975005961073806735733604454495675614232575
ADD
"
`;

exports[`assembly-parser should parse big octal number 1`] = `
"PUSHINT_LONG 794889263257962974796277498092801308291525640763748664903194643469338087775424965801409745320266996710649718116931109481559848982586784968419475084821084743272680947722675151641735826243378403750534655587182832000457137589153821622271
PUSHINT_LONG -794889263257962974796277498092801308291525640763748664903194643469338087775424965801409745320266996710649718116931109481559848982586784968419475084821084743272680947722675151641735826243378403750534655587182832000457137589153821622271
ADD
"
`;

exports[`assembly-parser should parse binary number 1`] = `
"PUSHINT_16 1023
PUSHINT_16 -1023
ADD
"
`;

exports[`assembly-parser should parse binary number with _ 1`] = `
"PUSHINT_16 1023
PUSHINT_16 -1023
ADD
"
`;

exports[`assembly-parser should parse decimal number 1`] = `
"PUSHINT_16 999999
PUSHINT_16 -999999
ADD
"
`;

exports[`assembly-parser should parse decimal number with _ 1`] = `
"PUSHINT_16 999999
PUSHINT_16 -999999
ADD
"
`;

exports[`assembly-parser should parse hex number 1`] = `
"PUSHINT_16 4095
PUSHINT_16 -4095
ADD
"
`;

exports[`assembly-parser should parse hex number with _ 1`] = `
"PUSHINT_16 4095
PUSHINT_16 -4095
ADD
"
`;

exports[`assembly-parser should parse octal number 1`] = `
"PUSHINT_16 32767
PUSHINT_16 -32767
ADD
"
`;

exports[`assembly-parser should parse octal number with _ 1`] = `
"PUSHINT_16 32767
PUSHINT_16 -32767
ADD
"
`;

exports[`assembly-parser should parse simple assembly 1`] = `
"PUSHINT_4 10
PUSHINT_4 5
ADD
"
`;
